### ISSUE TITLE
Steam group chats no longer work?

### ISSUE THREAD

**Issue by ItsEddie** [SteamRE/SteamKit#408]
I've been registering  SteamFriends.ChatMsgCallback but it never seems to fire. All other functionality does seem to work - it's just the group ones. Strangely, it does seem to recognize when I go in and out of a group chat.
```



        static void OnGroupMessaged(SteamFriends.ChatMsgCallback obj)
        {
            throw new NotImplementedException();
        }

      CallbackManager.Subscribe < SteamFriends.ChatMsgCallback>(SteamBotChat.OnGroupMessaged);


```

**Comment by JustArchi** [comment]
Sounds like chat ghost to me - typical chat room restart in which nobody is in chat anymore.

There is no good way to fight with it either, most stupid one is simply joining the chat on usual basis, such as each hour. A bit better one is doing the same but only when you don't register any activity such as chat message or join/leave, this could indicate that chat ghosted as well.

The best one that I use right now is tracking number of active users internally and checking number of chat users on steam community group page. If it's suddenly less than the amount that bot registered then very likely this is a ghost as well. Works decent.

**Comment by ItsEddie** [comment]
Any ideas on how to go about this more practically? I'm trying to bot a few dozen group chats and process what's being said. I'm pretty new to Steamkit and bots in general, used to work with making games.

**Comment by ItsEddie** [comment]
Any examples would be sweet too.

**Comment by Netshroud** [comment]
@JustArchi any idea what the Steam client does to prevent this? Or does it suffer from this as well?

**Comment by ItsEddie** [comment]
I'm surprised Steam does even anything. I've been logging into the same account every two minutes for debugging, got it also logged in on another computer, and haven't gotten any issues at all. This is really a great library.

**Comment by JustArchi** [comment]
@ItsEddie Most practical yet very stupid approach is like I said, registering last chat event for each chatID. For example `DateTime LastChatEvent`. You set it to `DateTime.UtcNow` on `ChatActionResultCallback`, `ChatEnterCallback`, `ChatMemberInfoCallback`, `ChatMsgCallback` and `ChatRoomInfoCallback` that arrives for given `chatID`, you have ID in each callback.

Each chat has a `Timer` that triggers each minute. If `DateTime.UtcNow - LastChatEvent > TimeSpan.FromMinutes(1)` (or anything other you set as "might ghost due to lack of activity" you trigger `SteamFriends.JoinChat(chatID)` that will also result in setting it back to `DateTime.UtcNow` because it triggers `ChatEnterCallback`. This is the best I can come up with **without** bringing in HTTP check.

@Netshroud Absolutely nothing. If chat ghosts, there are no messages happening, nobody is leaving/entering, and the only "indicator" that chat ghost happened is when one of chat visitors go offline - then you have a nice fuckup of user being on chat with "last online: 5 minutes ago" or something like that.

Basically I added very "stupid" way of detecting chat ghosts - like pointed out above, so when chat ghost happens my bot is messaging all chat visitors about ghost, then reconnects to chat itself. This is working just fine and allowed me to fight with this almost perfectly, since the worst case is bot telling you "chat ghosted, close tab then click here to reconnect: steamfriends://..."

**Comment by JustArchi** [comment]
And BTW this issue is driving me crazy as well since from SK2 there is absolutely no way of detecting that whatsoever, which is caused by Steam client not being aware of it at all. Sure you can "ignore" this problem by doing `SteamFriends.JoinChat(chatID)` on usual basis (e.g. like I suggested above, each minute) but detecting the problem is far more complicated and only possible by comparing some external resource (in my case, group page stating number of chat users) with what bot actually knows (number of chat visitors calculated from initial chat members list callback + tracking chat actions such as entering/leaving).

**Comment by ItsEddie** [comment]
Your approach is exactly how I created a spam filter, well almost the same.

`                Console.WriteLine(SteamBot.SteamFriends.GetFriendPersonaName(ID) + " said: " + Message);
                LogOther(ID, callback.Message);

                if (!SpamProtection.ContainsKey(ID))
                {
                    LastMessages[ID] = Message;
                    SpamProtection[ID] = DateTime.Now;
                    ChatSessions[ID] = ChatFactory.Create(ChatterBotType.PANDORABOTS,
                        "b0dafd24ee35a477").CreateSession();

                    new Thread(() => ThinkofResponse(ID, Message)).Start();
                }
                else if ((DateTime.Now - SpamProtection[ID]).TotalMilliseconds > 600)
                {
                    LastMessages[ID] = Message;

                    new Thread(() => ThinkofResponse(ID, Message)).Start();
                }
`

Anyways, I manually go into chatrooms, maybe that was part of my problem too. I'll take a look into it later, thank you a lot for your quick and kind responses.

By the way, something else - do you have a way to interact with the Steam community market? Or is that not yet available, or not even part of this kit?

I'm talking just getting prices for certain items, maybe buying ones, and selling some from the inventory.

**Comment by JustArchi** [comment]
Steam community, store as well as any HTTP-based functionality is out of the scope of SK2. You can find some functions extracted and implemented in my ASF project, but nothing related to market, as botting market is against Steam ToS, so I'm not dealing with that.

> Anyways, I manually go into chatrooms, maybe that was part of my problem too.

Could be, Steam network doesn't work well when there are multiple sessions with the same `LoginID`. This is mainly why I added #217 - you should make use of it if you're not doing that yet.

**Comment by ItsEddie** [comment]
Cheers.

**Comment by ItsEddie** [comment]
Good news. The reason the method wouldn't be executed is indeed that I manually joined it. After turning off my secondary computer using SteamFriends to join the chat, it finally throws the unimplemented error.

**Comment by yaakov-h** [comment]
Closing as this isn't a bug in SteamKit.
